### ISSUE TITLE
Potential information leakage in link prediction

### ISSUE THREAD

**Issue by geon0325** [snap-stanford/relbench#288]
Hello! I have a question regarding the potential information leakage during evaluations in link prediction. 

In **examples/idgnn_link.py** (and **examples/gnn_link.py**), dataloaders for train, validation, and test splits are defined as:

```
for split in ["train", "val", "test"]:
    table = task.get_table(split)
    table_input = get_link_train_table_input(table, task)
    dst_nodes_dict[split] = table_input.dst_nodes
    loader_dict[split] = NeighborLoader(
        data,
        num_neighbors=num_neighbors,
        time_attr="time",
        input_nodes=table_input.src_nodes,
        input_time=table_input.src_time,
        subgraph_type="bidirectional",
        batch_size=args.batch_size,
        temporal_strategy=args.temporal_strategy,
        shuffle=split == "train",
        num_workers=args.num_workers,
        persistent_workers=args.num_workers > 0,
    )
```

Here is a concern: Since **data** includes all train, validation, and test edges, for each *src_node*, its neighbors from all splits (train + validation + test) will be retrieved. This could lead to information leakage during evaluation as test (validation) edges are also iterated during test (validation) evaluation.

This issue is particularly important in **examples/idgnn_link.py** because, during evaluation, scores are computed only for the neighbors sampled from the dataloaders, while scores for the remaining nodes are set to zero. Thus, if neighbors connected by test edges (which should not be sampled at the test phase) are included, they may be unfairly prioritized over other nodes.

```
@torch.no_grad()
def test(loader: NeighborLoader) -> np.ndarray:
    model.eval()

    pred_list: list[Tensor] = []
    for batch in tqdm(loader):
        batch = batch.to(device)
        out = (
            model.forward_dst_readout(
                batch, task.src_entity_table, task.dst_entity_table
            )
            .detach()
            .flatten()
        )
        batch_size = batch[task.src_entity_table].batch_size
        scores = torch.zeros(batch_size, task.num_dst_nodes, device=out.device)
        scores[
            batch[task.dst_entity_table].batch, batch[task.dst_entity_table].n_id
        ] = torch.sigmoid(out)
        _, pred_mini = torch.topk(scores, k=task.eval_k, dim=1)
        pred_list.append(pred_mini)
    pred = torch.cat(pred_list, dim=0).cpu().numpy()
    return pred
```

To address this issue, I was wondering if we need to modify the **data** used for defining dataloaders to include only train edges for the validation dataloader and include train + validation edges for the test dataloader.

Thank you for your attention to this matter in advance!

**Comment by geon0325** [comment]
For reference: Even when the training phase is commented out (thus the model is not trained), the evaluation metrics yielded a MAP of 1.41 on the validation set and 1.50 on the test set in the **rel-hm** dataset. These results (without models trained) outperform the performance of GraphSAGE and Past Visit reported in the paper.

```
state_dict = None
best_val_metric = 0
for epoch in range(1, args.epochs + 1):
    #train_loss = train()
    if epoch % args.eval_epochs_interval == 0:
        val_pred = test(loader_dict["val"])
        val_metrics = task.evaluate(val_pred, task.get_table("val"))

        test_pred = test(loader_dict["test"])
        test_metrics = task.evaluate(test_pred)

        print(f"Val metrics: {val_metrics}")
        print(f"Test metrics: {test_metrics}")

        if val_metrics[tune_metric] > best_val_metric:
            best_val_metric = val_metrics[tune_metric]
            state_dict = copy.deepcopy(model.state_dict())
```

**Comment by rishabh-ranjan** [comment]
Thanks for bringing this up!

@joshr17 what do you think?

**Comment by geon0325** [comment]
Thank you for taking my issue into account! Let me clarify my concerns further:

Let’s consider the **rel-hm** dataset as an example. Initially, we load the following data:
```
dataset: Dataset = get_dataset('rel-hm', download=False)
task: RecommendationTask = get_task('rel-hm', "user-item-purchase", download=False)

tables = {}
for split in ["train", "val", "test"]:
    tables[split] = task.get_table(split)
```

Here, the `dataset` object is used to generate the graph using the `make_pkey_fkey_graph` function from `relbench.modeling.graph`. Specifically, the `transactions` dataframe within dataset is utilized. For instance, in **rel-hm**, this includes transactions ranging **from 2019-09-07 to 2020-09-14**, from:
```
min(dataset.get_db().table_dict['transactions'].df['t_dat'])
max(dataset.get_db().table_dict['transactions'].df['t_dat'])
```

However, this appears to include edges from the validation and test sets. When examining the date ranges for each split:
```
min(tables["train"].df['timestamp']), max(tables["train"].df['timestamp'])
min(tables["val"].df['timestamp']), max(tables["val"].df['timestamp'])
min(tables["test"].df['timestamp']), max(tables["test"].df['timestamp'])
```

- Train set: **2019-09-09 to 2020-08-31**
- Validation set: **2020-09-07** (single day)
- Test set: **2020-09-14** (single day)

Since I couldn’t find any part of `make_pkey_fkey_graph` that filters out edges from the validation and test sets, I’m concerned that future transactions (from validation and test sets) might be included in the graph. This could mean they are used during both the training and evaluation phases.

**Comment by quang-truong** [comment]
Hi @geon0325, shouldn't `NeighborLoader` be able to handle temporal sampling as `time_attr` and `input_time` are set?

**Comment by geon0325** [comment]
Thank you, @quang-truong for pointing this out! I think `input_time` does prevent future edges from being included as neighbors. 

One minor concern is that the timestamps of `dataset.get_db().table_dict['transactions'].df['t_dat']` which is used for constructing the graph and `task.get_table("train")` which is used for generating `input_time` are not aligned. Whether this misalignment affects information leakage requires further examination. I will close this one and open a new issue if needed. 

Thanks!
